### PR TITLE
Add ability to collect system information from users' systems

### LINKED ISSUE #10662: Use the battlefield news query to send system stats to the master server
Before we can fix #10661 we need an idea of how many players this will affect.
The battlefield news query runs once per game start, so it gives us a simple way to track active player statistics (unless the player explicitly opts out).

Implementation:

1. Add a `http://master.openra.net/gamenews` page, which can initially just do a http redirect to http://openra.net/gamenews.
2.  Change the `NewsUrl` to this new value.
2. Add an advanced setting for sending system statistics (enabled by default).
3. If the setting is enabled, send the current mod, `Platform.CurrentPlatform`, `Environment.OSVersion`, and `Platform.RuntimeVersion` (these are the same things we include in exception.log) as either GET or POST data along with the news query.
4. (can be left until after release) Change the gamenews page to start logging this information so that we look at aggregate OS/usage stats.

If we want to do anything about #10661 in the next year we really need to get this logging in the next release, so adding to the milestone.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30385358-use-the-battlefield-news-query-to-send-system-stats-to-the-master-server?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by obrakmann**
Closes #10662.

Edit: Just to be sure: DON'T MERGE BEFORE THE REDIRECT ON master.openra.net IS SET UP!

### CODE REVIEW

**Comment by obrakmann** [comment]
Added version of the 'all' mod to the query.

File: OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs [hunk 1]
```diff
@@ -251,6 +251,21 @@ void LoadMapIntoEditor(string uid)
...
+			if (!Game.Settings.Game.PlayerCensusParticipation)
+				return null;
+
+			return "?id={0}&platform={1}&os={2}&runtime={3}&lang={4}&version={5}&mod={6}".F(
+				Uri.EscapeUriString(Game.Settings.Game.UUID),
+				Uri.EscapeUriString(Platform.CurrentPlatform.ToString()),
+				Uri.EscapeUriString(Environment.OSVersion.ToString()),
+				Uri.EscapeUriString(Platform.RuntimeVersion),
+				Uri.EscapeUriString(System.Globalization.CultureInfo.InstalledUICulture.TwoLetterISOLanguageName),
+				Uri.EscapeUriString(ModMetadata.AllMods["all"].Version),
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs]: Can we log the current mod's ID and version instead?  IMO this is more useful than the "all" version and the mod title.

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs]: or we could keep this, but replace the mod title with ^

Comment (obrakmann) [inline OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs]: No, that's fine with me. I was actually looking for the id, but totally missed that field. Updated.

**Comment by pchote** [comment]
As a starting point (so this can be merged) the redirect can be done by dropping a `gamenews` file on the master server with the following contents (may contain typos - it's been a long time since I last wrote php):
```php
<?php
header('Location: http://www.openra.net/gamenews');
?>
```

**Comment by obrakmann** [comment]
Seperated game version and mod version.

File: mods/cnc/chrome/settings.yaml [hunk 2]
```diff
@@ -508,43 +515,43 @@ Container@SETTINGS_PANEL:
...
 							Text: Fetch Community News
-						Checkbox@PERFGRAPH_CHECKBOX:
+						Checkbox@CENSUS_CHECKBOX:
 							X: 310
 							Y: 70
 							Width: 300
 							Height: 20
 							Font: Regular
-							Text: Show Performance Graph
+							Text: Participate in Player Census
```
Comment (Phrohdoh) [inline mods/cnc/chrome/settings.yaml]: Could we be a bit more straightforward and call this (and other necessary places)`Send System Information` instead?

**Comment by obrakmann** [comment]
Changed the wording

**Comment by pchote** [comment]
Can we pretty please also include `glGetString(GL_VERSION)` in this data?  We will be able to dramatically improve our rendering code if/when we can enforce GL >= 3

**Comment by obrakmann** [comment]
Sure. You'll need to tell me how I should massacre the code to allow me to get at that, though. The OpenGL class is `internal static` and OpenRA.Mods.Common doesn't currently have access to the Platforms assemblies.

**Comment by pchote** [comment]
I suggest exposing a `string OpenGLVersion` property on Game.Renderer.

**Comment by obrakmann** [comment]
Sorry, but that doesn't help. I think it would be quicker if you sent a patch.

**Comment by pchote** [comment]
ok

**Comment by LipkeGu** [comment]
We should also give the User the ability to disable the collection, and inform them.

**Comment by pchote** [comment]
I've set up the redirect, and will try and send you a patch for the GL version tonight.

File: mods/cnc/chrome/settings.yaml [hunk 3]
```diff
@@ -508,43 +515,43 @@ Container@SETTINGS_PANEL:
...
 							Text: Fetch Community News
-						Checkbox@PERFGRAPH_CHECKBOX:
+						Checkbox@SENDSYSINFO_CHECKBOX:
 							X: 310
 							Y: 70
 							Width: 300
 							Height: 20
 							Font: Regular
-							Text: Show Performance Graph
+							Text: Send system information
```
Comment (pchote) [inline mods/cnc/chrome/settings.yaml]: System and information should be capitalized to match the other options.

File: mods/ra/chrome/settings.yaml [hunk 4]
```diff
@@ -509,43 +516,43 @@ Background@SETTINGS_PANEL:
...
 					Text: Fetch Community News
-				Checkbox@PERFGRAPH_CHECKBOX:
+				Checkbox@SENDSYSINFO_CHECKBOX:
 					X: 310
 					Y: 70
 					Width: 300
 					Height: 20
 					Font: Regular
-					Text: Show Performance Graph
+					Text: Send system information
```
Comment (pchote) [inline mods/ra/chrome/settings.yaml]: Here too

File: mods/cnc/chrome/settings.yaml (hunk 3 above)
Comment (pchote) [inline mods/cnc/chrome/settings.yaml]: Can you put, in the same `Tiny` font that we use for the video desc label (probably needs to be split over two lines) something like: "Your OS, OpenGL version, and and .NET runtime version, and language settings are sent with an anonymous ID to help priorise future development"?

**Comment by pchote** [comment]
https://github.com/obrakmann/OpenRA/pull/5

**Comment by pchote** [comment]
Code looks good, and only a couple of minor UI suggestions.  This gets a deserved :+1: after fixing those and merging the GL patch.

Sample query URL: http://master.openra.net/gamenews?id=3fd3b08c-7933-47e9-a5b2-52e6bbceeb5b&platform=OSX&os=Unix%2015.3.0.0&runtime=Mono%204.2.2%20(explicit/996df3c%20Wed%20Jan%2020%2000:19:48%20EST%202016)%20CLR%204.0.30319.17020&gl=2.1%20INTEL-10.12.13lang=en&version=%7BDEV_VERSION%7D&mod=cnc&modversion=%7BDEV_VERSION%7D

which decodes to:
```
http://master.openra.net/gamenews
   ?id=3fd3b08c-7933-47e9-a5b2-52e6bbceeb5b&platform=OSX
   &os=Unix 15.3.0.0
   &runtime=Mono 4.2.2 (explicit/996df3c Wed Jan 20 00:19:48 EST 2016) CLR 4.0.30319.17020
   &gl=2.1 INTEL-10.12.13
   &lang=en
   &version={DEV_VERSION}
   &mod=cnc
   &modversion={DEV_VERSION}
```

**Comment by obrakmann** [comment]
Updated.
screenshot from 2016-03-02 21 27 00

**Comment by Phrohdoh** [comment]
Off topic but this just showcases our horrible font rendering. :(

**Comment by abcdefg30** [comment]
:+1:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ad9f6bc6b72594194c9511b172c004d5457e2900)
I put it under `Engine / Modding` now (although @penev92 suggested to put it under `Spyware`).

Might be worth to mention that somewhere in the newspost, so that people who don't want to give their data don't rage later on (hence the `Spyware` above ^^)?

**Comment by pchote** [comment]
I was already planning to mention it in the news, yeah.

File: mods/cnc/chrome/settings.yaml [hunk 5]
```diff
@@ -530,43 +537,51 @@ Container@SETTINGS_PANEL:
...
-							Text: Show Performance Graph
+							Text: Send System Information
+						Label@SENDSYSINFO_DESC:
+							X: 310
+							Y: 85
+							Width: 250
+							Height: 30
+							Font: Tiny
+							WordWrap: True
+							Text: Your Operating System, OpenGL and .NET runtime versions, and language settings will be sent along with an anonymous ID to help priorise future development.
```
Comment (ScottNZ) [inline mods/cnc/chrome/settings.yaml]: `priorise` should have been `prioritize`

**Pull request merged**